batcheval: absorb command registration

tbg opened the pull request:
You'll notice that `EndTransaction` is still lingering in `replica_command.go`,
but that will be done in a follow-up.

Touches #18779.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/20329)
<!-- Reviewable:end -->

nvanbenschoten commented:
This cleanup is really satisfying to follow!

---

Reviewed 38 of 38 files at r1.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[pkg/storage/cclglue.go, line 25 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20329#-L-8LC8b7Jc9BroGg-2I:-L-8LC8b7Jc9BroGg-2J:b-jjtdkx) ([raw file](https://github.com/cockroachdb/cockroach/blob/2651d00b14bc58a95d8047b8ad3f6c63663a1166/pkg/storage/cclglue.go#L25)):*
> ```Go
> )
> 
> // TODO(tschottdorf): seems straightforward enough to shoehorn this into
> ```

Why not handle this here as well?

---

*[pkg/storage/helpers_test.go, line 474 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20329#-L-8LRA9oU-qFLEgUeoP:-L-8LRA9oU-qFLEgUeoQ:b9r8wsj) ([raw file](https://github.com/cockroachdb/cockroach/blob/2651d00b14bc58a95d8047b8ad3f6c63663a1166/pkg/storage/helpers_test.go#L474)):*
> ```Go
> 	}
> 
> 	batcheval.RegisterCommand(roachpb.AddSSTable, nil, nil)
> ```

~Why are we doing this twice?~

Never mind, but I'd still push for an `UnregisterCommand` function.

---

*[pkg/storage/replica_command.go, line 160 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20329#-L-8LjNagEHepm-te7AU:-L-8LjNagEHepm-te7AV:b-m2c3d1) ([raw file](https://github.com/cockroachdb/cockroach/blob/2651d00b14bc58a95d8047b8ad3f6c63663a1166/pkg/storage/replica_command.go#L160)):*
> ```Go
> 
> func init() {
> 	batcheval.RegisterCommand(roachpb.EndTransaction, declareKeysEndTransaction, evalEndTransaction)
> ```

I know you're going to get to this soon, but could you add a TODO here?

---

*[pkg/storage/batcheval/cmd_begin_transaction.go, line 34 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20329#-L-8MHsT2MP-6CO1BJX-:-L-8MHsT2MP-6CO1BJX0:blcbvpj) ([raw file](https://github.com/cockroachdb/cockroach/blob/2651d00b14bc58a95d8047b8ad3f6c63663a1166/pkg/storage/batcheval/cmd_begin_transaction.go#L34)):*
> ```Go
> }
> 
> // DeclareKeysWriteTransaction is the shared portion of
> ```

Consider a TODO to unexport this when EndTxn is moved into `batcheval`.

---

*[pkg/storage/batcheval/cmd_conditional_put.go, line 29 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20329#-L-8NEaWodKtxFGSEMF7:-L-8NEaWodKtxFGSEMF8:bq3s930) ([raw file](https://github.com/cockroachdb/cockroach/blob/2651d00b14bc58a95d8047b8ad3f6c63663a1166/pkg/storage/batcheval/cmd_conditional_put.go#L29)):*
> ```Go
> }
> 
> // ConditionalPut sets the value for a specified key only if
> ```

Do these functions need to be exported?

---

*[pkg/storage/batcheval/cmd_deprecated_verify_checksum.go, line 26 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20329#-L-8NYcE72Yql4yhmTwk:-L-8NYcE72Yql4yhmTwl:bae93tb) ([raw file](https://github.com/cockroachdb/cockroach/blob/2651d00b14bc58a95d8047b8ad3f6c63663a1166/pkg/storage/batcheval/cmd_deprecated_verify_checksum.go#L26)):*
> ```Go
> 
> func init() {
> 	RegisterCommand(roachpb.DeprecatedVerifyChecksum, DefaultDeclareKeys, deprecatedVerifyChecksum)
> ```

nit: Why declare any keys for this command?

---

*[pkg/storage/batcheval/cmd_lease.go, line 32 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20329#-L-8NnaSj4ilfq1yoIfL:-L-8NnaSj4ilfq1yoIfM:btifqsm) ([raw file](https://github.com/cockroachdb/cockroach/blob/2651d00b14bc58a95d8047b8ad3f6c63663a1166/pkg/storage/batcheval/cmd_lease.go#L32)):*
> ```Go
> 
> func declareKeysRequestLease(
> 	desc roachpb.RangeDescriptor, header roachpb.Header, req roachpb.Request, spans *spanset.SpanSet,
> ```

The linter might yell at you about this long line.

---

*[pkg/storage/batcheval/cmd_resolve_intent_test.go, line 36 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20329#-L-8O48xNdooHfaTM-dj:-L-8O48xNdooHfaTM-dk:bo7rqpc) ([raw file](https://github.com/cockroachdb/cockroach/blob/2651d00b14bc58a95d8047b8ad3f6c63663a1166/pkg/storage/batcheval/cmd_resolve_intent_test.go#L36)):*
> ```Go
> 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
> 	"github.com/cockroachdb/cockroach/pkg/util/uuid"
> 	opentracing "github.com/opentracing/opentracing-go"
> ```

nit: move up

---

*[pkg/storage/batcheval/cmd_scan.go, line 30 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20329#-L-8OTWs9bcw_zOL1QPu:-L-8OTWs9bcw_zOL1QPv:b-ped1rn) ([raw file](https://github.com/cockroachdb/cockroach/blob/2651d00b14bc58a95d8047b8ad3f6c63663a1166/pkg/storage/batcheval/cmd_scan.go#L30)):*
> ```Go
> type Command struct {
> 	// DeclareKeys adds all keys this command touches to the given spanSet.
> 	DeclareKeys func(roachpb.RangeDescriptor, roachpb.Header, roachpb.Request, *spanset.SpanSet)
> ```

Consider creating typedefs of these two functions and sharing between here and `RegisterCommands` to avoid duplication.

---

*[pkg/storage/batcheval/cmd_scan.go, line 52 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20329#-L-8Oc2X_aBdJSrtNmyx:-L-8Oc2X_aBdJSrtNmyy:b-uddafs) ([raw file](https://github.com/cockroachdb/cockroach/blob/2651d00b14bc58a95d8047b8ad3f6c63663a1166/pkg/storage/batcheval/cmd_scan.go#L52)):*
> ```Go
> 	impl func(context.Context, engine.ReadWriter, CommandArgs, roachpb.Response) (result.Result, error),
> ) {
> 	if declare == nil && impl == nil {
> ```

Why not just create a `UnregisterCommand` function? Seems less obscure.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20329#-:-L-8P71VEeHbSoIlu_Co:bfo3epg)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[pkg/storage/cclglue.go, line 25 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20329#-L-8LC8b7Jc9BroGg-2I:-L-Cdqv7eGMu5nxUuADx:bivx5ci) ([raw file](https://github.com/cockroachdb/cockroach/blob/2651d00b14bc58a95d8047b8ad3f6c63663a1166/pkg/storage/cclglue.go#L25)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Why not handle this here as well?
</blockquote></details>

I take it back, `importCmd` is an admin command. Removed the TODO.

---

*[pkg/storage/helpers_test.go, line 474 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20329#-L-8LRA9oU-qFLEgUeoP:-L-CbjcuFQZuFMZVV9uJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2651d00b14bc58a95d8047b8ad3f6c63663a1166/pkg/storage/helpers_test.go#L474)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

~Why are we doing this twice?~

Never mind, but I'd still push for an `UnregisterCommand` function.
</blockquote></details>

Done.

---

*[pkg/storage/replica_command.go, line 160 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20329#-L-8LjNagEHepm-te7AU:-L-Cbs_VU5tuUC2bBS7H:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2651d00b14bc58a95d8047b8ad3f6c63663a1166/pkg/storage/replica_command.go#L160)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I know you're going to get to this soon, but could you add a TODO here?
</blockquote></details>

Done.

---

*[pkg/storage/batcheval/cmd_begin_transaction.go, line 34 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20329#-L-8MHsT2MP-6CO1BJX-:-L-CbtQOXXjvK9ieU_gB:b8u1dl6) ([raw file](https://github.com/cockroachdb/cockroach/blob/2651d00b14bc58a95d8047b8ad3f6c63663a1166/pkg/storage/batcheval/cmd_begin_transaction.go#L34)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Consider a TODO to unexport this when EndTxn is moved into `batcheval`.
</blockquote></details>

Done. (elsewhere)

---

*[pkg/storage/batcheval/cmd_conditional_put.go, line 29 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20329#-L-8NEaWodKtxFGSEMF7:-L-CbwSOLFESFWzJDwpw:bvovb8o) ([raw file](https://github.com/cockroachdb/cockroach/blob/2651d00b14bc58a95d8047b8ad3f6c63663a1166/pkg/storage/batcheval/cmd_conditional_put.go#L29)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Do these functions need to be exported?
</blockquote></details>

no, but I decided to leave them exported because they are the meat of what this package provides. We don't really spend a lot of time looking at godoc, but if we did, we'd want to see them and their comments.

---

*[pkg/storage/batcheval/cmd_deprecated_verify_checksum.go, line 26 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20329#-L-8NYcE72Yql4yhmTwk:-L-CcC9z89YWAuojhM5x:b6cq5z5) ([raw file](https://github.com/cockroachdb/cockroach/blob/2651d00b14bc58a95d8047b8ad3f6c63663a1166/pkg/storage/batcheval/cmd_deprecated_verify_checksum.go#L26)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: Why declare any keys for this command?
</blockquote></details>

You could take it a step further and ask "why not delete this command". We can probably delete this, but I'd rather not do it as a drive-by and have thus opted to preserve the existing code verbatim.

---

*[pkg/storage/batcheval/cmd_lease.go, line 32 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20329#-L-8NnaSj4ilfq1yoIfL:-L-CcK66xxO3F1Wsn27S:b-knwqqn) ([raw file](https://github.com/cockroachdb/cockroach/blob/2651d00b14bc58a95d8047b8ad3f6c63663a1166/pkg/storage/batcheval/cmd_lease.go#L32)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

The linter might yell at you about this long line.
</blockquote></details>

The linter and I are friends.

---

*[pkg/storage/batcheval/cmd_resolve_intent_test.go, line 36 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20329#-L-8O48xNdooHfaTM-dj:-L-CcfJ1dqFkuDn4jW6Q:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2651d00b14bc58a95d8047b8ad3f6c63663a1166/pkg/storage/batcheval/cmd_resolve_intent_test.go#L36)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: move up
</blockquote></details>

Done.

---

*[pkg/storage/batcheval/cmd_scan.go, line 30 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20329#-L-8OTWs9bcw_zOL1QPu:-L-CcnR79gv7xg8yjnog:b-go6ew4) ([raw file](https://github.com/cockroachdb/cockroach/blob/2651d00b14bc58a95d8047b8ad3f6c63663a1166/pkg/storage/batcheval/cmd_scan.go#L30)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Consider creating typedefs of these two functions and sharing between here and `RegisterCommands` to avoid duplication.
</blockquote></details>

I'm not a huge fan of typedefs for fns because all they usually do is add an extra hop to look at what you're actually supposed to pass. I'll leave as is in this instance.

---

*[pkg/storage/batcheval/cmd_scan.go, line 52 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20329#-L-8Oc2X_aBdJSrtNmyx:-L-Ccx48Mgcp_J7dIw4R:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2651d00b14bc58a95d8047b8ad3f6c63663a1166/pkg/storage/batcheval/cmd_scan.go#L52)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Why not just create a `UnregisterCommand` function? Seems less obscure.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20329)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

---

Reviewed 10 of 10 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20329#-:-L-CfmYBYWZoax16beST:b-uwkowd)*
<!-- Sent from Reviewable.io -->

tbg committed:
batcheval: absorb command registration

You'll notice that `EndTransaction` is still lingering in `replica_command.go`,
but that will be done in a follow-up.

Touches #18779.
